src/main/java/build/buf/protovalidate/internal/evaluator/EvaluatorBase.java

```diff
@@ -0,0 +1,61 @@
+// Copyright 2023-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package build.buf.protovalidate.internal.evaluator;
+
+import build.buf.protovalidate.internal.errors.FieldPathUtils;
+import build.buf.validate.FieldPath;
+import build.buf.validate.FieldPathElement;
+import java.util.ArrayList;
+import java.util.List;
+import javax.annotation.Nullable;
+
+class EvaluatorBase {
```

Comment:
Instead of extending this in every evaluator type, I think it would be better to use composition and maybe call it a `ConstraintViolationHelper` -- since it's really just used to get values for building a violation.

There's a lot of places that now call `super(evaluator)` that IMO don't look intuitive. Also, this doesn't provide any base `Evaluator` method implementations so it is awkward as a base class.

Reply:
Yeah, I was going to try to extract more common functionality but it didn't work out so it's a bit pointless now. Now it's `ConstraintViolationHelper` and not being used as a base class.